extensions/webpush.md

```diff
@@ -0,0 +1,96 @@
+---
+title: "Web Push Extension"
+layout: spec
+copyrights:
+  - name: "Simon Ser"
+    period: "2021"
+    email: "contact@emersion.fr"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the unprefixed `webpush` CAP name. Instead, implementations SHOULD use the `draft/webpush` CAP name to be interoperable with other software implementing a compatible work-in-progress version. The final version of the specification will use unprefixed CAP names.
+
+## Description
+
+Historically, IRC clients have relied on keeping a TCP connection alive to receive notifications about new events. However, this design has limitations:
+
+- It doesn't bode well with some platforms such as Android, iOS or the Web. On these platforms, the connection to the IRC server can be severed (e.g. when the IRC client isn't in the foreground), resulting in IRC events not received.
+- Battery-powered devices aim to avoid any unnecessary wake-up of the modem hardware. IRC connections don't make the difference between messages which may be important to the user (e.g. messages targeting the user directly) and the rest of the messages. As a result messages are frequently sent over the IRC connection, resulting in battery drain.
+
+To address these limitations, various push notification mechanisms have been designed. This specification standardizes an extension for Web Push.
+
+```
+      ┌────────────┐              ┌────────────┐
+      │            │  Subscribe   │            │
+      │            ├─────────────►│            │
+      │ IRC client │              │ IRC server │
+      │            │              │            │
+      │            │              │            │
+      └────────────┘              └─────┬──────┘
+             ▲                          │
+             │                          │
+        Push │                          │Push
+notification │                          │notification
+             │       ┌──────────┐       │
+             │       │          │       │
+             └───────┤ Web Push │◄──────┘
+                     │  Server  │
+                     │          │
+                     └──────────┘
+```
+
+Web Push is defined in [RFC 8030], [RFC 8291] and [RFC 8292].
+
+Although Web Push has been designed for the Web, it can be used on other platforms as well. Web Push provides a vendor-neutral standard to send push notifications.
+
+## Implementation
+
+The `webpush` capability allows clients to subscribe to Web Push and receive
+notifications for messages of interest.
```

Comment:
How does a user/client indicate which messages would be of interest? There doesn't appear to be mechanisms in-place in the currently offered commands to indicate which messages would be of interest.

Reply:
Yup, as discussed in #ircv3, this part isn't addressed yet. For now, it's spec'ed as a "server-defined subset of IRC messages".

I think there are multiple ways to go about this:

1. Keep as-is, let servers decide completely. Leave it to a future extension to let clients indicate messages of interest. This can be useful for chathistory as well, to avoid fetching all unread history when connecting.
2. Indicate a minimal set of messages to forward (e.g. PRIVMSG/NOTICE/INVITE with user as target, plus PRIVMSG/NOTICE highlights). Let servers send more if they want to. Clients can always ignore some push notifications.
3. Build a webpush-specific command to indicate the messages of interest. This is a slippery slope, because the rules can be pretty complicated. If we go down this road I'd rather keep it simple and have just simple word matching.

Reply:
I think (1) is the superior option here.